flink-filesystems/flink-gs-fs-hadoop/src/main/java/org/apache/flink/fs/gs/writer/GSRecoverableWriterCommitter.java

```diff
@@ -0,0 +1,189 @@
+
+        // clean up after successful commit
+        cleanupTemporaryBlobs();
+    }
+
+    @Override
+    public void commitAfterRecovery() throws IOException {
+
+        // see discussion: https://github.com/apache/flink/pull/15599#discussion_r623127365
+        // only write the final blob if it doesn't already exist
+        Optional<GSBlobStorage.BlobMetadata> blobMetadata =
+                storage.getMetadata(recoverable.finalBlobIdentifier);
+        if (!blobMetadata.isPresent()) {
+            writeFinalBlob();
+        }
+
+        // clean up after successful commit
+        cleanupTemporaryBlobs();
+    }
+
+    @Override
+    public RecoverableWriter.CommitRecoverable getRecoverable() {
+        return recoverable;
+    }
+
+    /**
+     * Helper to compose an arbitrary number of blobs into a final blob, staying under the
+     * COMPOSE_MAX_BLOBS limit for any individual compose operation.
+     *
+     * @param sourceBlobIdentifiers The source blob ids to compose
+     * @param targetBlobIdentifier The target blob id for the composed result
+     */
+    private void composeBlobs(
+            List<GSBlobIdentifier> sourceBlobIdentifiers, GSBlobIdentifier targetBlobIdentifier) {
+        Preconditions.checkNotNull(sourceBlobIdentifiers);
+        Preconditions.checkArgument(sourceBlobIdentifiers.size() > 0);
+        Preconditions.checkNotNull(targetBlobIdentifier);
+
+        // split the source list into two parts; first, the ones we can compose in this operation
+        // (up to COMPOSE_MAX_BLOBS), and, second, whichever blobs are left over
+        final int composeToIndex =
+                Math.min(BlobUtils.COMPOSE_MAX_BLOBS, sourceBlobIdentifiers.size());
+        List<GSBlobIdentifier> composeBlobIds = sourceBlobIdentifiers.subList(0, composeToIndex);
+        List<GSBlobIdentifier> remainingBlobIds =
+                sourceBlobIdentifiers.subList(composeToIndex, sourceBlobIdentifiers.size());
+
+        // determine the resulting blob id for this compose operation. if this is the last compose,
+        // i.e. if there are no remaining blob ids, then the composed blob id is the originally
+        // specified target blob id. otherwise, we must create an intermediate blob id to hold the
+        // result of this compose operation
+        GSBlobIdentifier composedBlobId =
+                remainingBlobIds.isEmpty()
+                        ? targetBlobIdentifier
+                        : BlobUtils.generateTemporaryBlobIdentifier(
+                                recoverable.finalBlobIdentifier, options);
+
+        // compose the blobs
+        storage.compose(composeBlobIds, composedBlobId);
+
+        // if we have remaining blobs, add the composed blob id to the beginning of the list
+        // of remaining blob ids, and recurse
+        if (!remainingBlobIds.isEmpty()) {
+            remainingBlobIds.add(0, composedBlobId);
+            composeBlobs(remainingBlobIds, targetBlobIdentifier);
+        }
+    }
```

Comment:
@xintongsong I do see what you're saying about ```COMPOSE_MAX_BLOBS``` being a limitation of GCS blob storage and not necessarily of the abstract ```GSBlobStorage``` interface, and so this logic could, in principle, be moved to be internal to ```GSBlobStorageImpl```.

However, from a unit-testing perspective, I'd been hoping to keep the implementation of ```GSBlobStorageImpl``` as simple as possible, i.e. making each method's implementation as close to simple pass-through calls to the underlying Google storage interfaces as possible. In particular, I'm hoping to avoid putting any complicated logic in need of testing there. This is because the underlying Google storage interfaces are tricky to mock -- i.e. they have lots of methods and they sometimes return types that are themselves difficult to create or mock -- and so writing tests at that level will, in general, be more difficult.

Alternately, if I were to keep the ```COMPOSE_MAX_BLOBS```-aware compose logic outside of ```GSBlobStorageImpl```, then I could test that logic using a mocked implementation of ```GSBlobStorage``` -- i.e. ```MockBlobStorage``` -- which is easy to test against.

What would you think of an alternate approach, where I would move the ```COMPOSE_MAX_BLOBS```-aware compose logic out of ```GSRecoverableWriterCommitter#composeBlobs``` and into a utility method, say ```BlobUtils#composeBlobs```? That utility method would take the same parameters as the existing one plus: 
* A ```GSBlobStorage``` instance to operate against
* A parameter indicating the maximum number of blobs to compose at once

So, essentially, this new utility method would be a generic implementation of the "compose blobs with a limit on the number of blobs that can be composed in one step" algorithm, which I could test easily against MockBlobStorage. Then, ```GSRecoverableWriterCommitter``` could call that utility method, passing COMPOSE_MAX_BLOBS for the max number of blobs to compose at once.